ADR0003: where to develop TUF 1.0.0

sechkova opened the pull request:
Fixes #1126 

**Description of the changes being introduced by the pull request**:

Document the outcome of #1126: develop TUF 1.0.0 in a subdirectory of the current TUF implementation.

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

sechkova committed:
ADR0003: where to develop TUF 1.0.0

Document the outcome of #1126 to develop TUF 1.0.0
in a subdirectory of the current TUF implementation.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

lukpueh commented on docs/adr/0003-where-to-develop-TUF-1-0-0.md:
```diff
@@ -0,0 +1,46 @@
+Chosen option: "Develop TUF 1.0.0 in a subdirectory of the current TUF
+implementation", because we want to add the new TUF code gradually
+while keep maintaining the current implementation given limited
+maintenance resources.
```
Minor nit: Should we somehow underline that this option seems to have the least maintenance overhead compared to option 1 and 2, while allowing us to continue making releases with the old code unlike option 3?

lukpueh reviewed (approved):
Thanks for the PR, @sechkova. I have one minor nit (see inline) but other than that it looks solid.

lukpueh commented:
Not sure if this should be part of AD0003, or should have been part of [ADR0002](https://github.com/theupdateframework/tuf/blob/develop/docs/adr/0002-pre-1-0-deprecation-strategy.md), or neither, but we also need to decide how to transition from TUF 1.0.0 in a subdirectory to stand-alone TUF 1.0.0. Im fine if we cross that bridge when we get there (e.g. with the help of `git filter-branch`-sorcery). cc @joshuagl

trishankatdatadog commented:
I'm ok with developing in a separate dir in the current default branch, as long as we're not committed to legacy code. Should we clarify this in the ADR, if we all agree?

joshuagl commented:
> Not sure if this should be part of AD0003, or should have been part of [ADR0002](https://github.com/theupdateframework/tuf/blob/develop/docs/adr/0002-pre-1-0-deprecation-strategy.md), or neither, but we also need to decide how to transition from TUF 1.0.0 in a subdirectory to stand-alone TUF 1.0.0. Im fine if we cross that bridge when we get there (e.g. with the help of `git filter-branch`-sorcery). cc @joshuagl

The way I've been imaginging it, filter-branch wizardry would not be required. In my mind we'd have something like:
* flesh out tuf/api/*
* implement tuf/client/new-updater.py
* implement tuf/repository/*
* \<iterate\> 💻 
* git mv tuf/client/new-updater.py tuf/client/updater.py
* git rm tuf/*.py
* tag 1.0.0

joshuagl reviewed (approved):
Thanks for writing this up @sechkova !

joshuagl commented:
> I'm ok with developing in a separate dir in the current default branch, as long as we're not committed to legacy code. Should we clarify this in the ADR, if we all agree?

I believe we all agree. I think it's implicit in ADR 0002 that we are not committed to legacy code?

joshuagl commented:
Per https://github.com/theupdateframework/tuf/pull/1203#pullrequestreview-535266482 and the discussion in #1127 we still need to document:

> a deprecation/API stability strategy for releases after the refactor

Perhaps we can ensure we are not committed to maintaining legacy code with that document?

lukpueh commented:
> The way I've been imaginging it, filter-branch wizardry would not be required. In my mind we'd have something like:
> 
> * flesh out tuf/api/*
> * implement tuf/client/new-updater.py
> * implement tuf/repository/*
> * <iterate> 💻
> * git mv tuf/client/new-updater.py tuf/client/updater.py
> * git rm tuf/*.py
> * tag 1.0.0

Sounds like a good plan and a good fit for an addendum to ADR0003?

joshuagl commented:
> > The way I've been imaginging it, filter-branch wizardry would not be required. In my mind we'd have something like:
> > 
> > * flesh out tuf/api/*
> > * implement tuf/client/new-updater.py
> > * implement tuf/repository/*
> > * 💻
> > * git mv tuf/client/new-updater.py tuf/client/updater.py
> > * git rm tuf/*.py
> > * tag 1.0.0
> 
> Sounds like a good plan and a good fit for an addendum to ADR0003?

Agreed. Would you mind outlining the above in ADR0003 @sechkova ?

sechkova committed:
ADR003: describe transition to stand-alone TUF

Describe the steps for transitioning from TUF 1.0.0
in a subdirectory to stand-alone TUF 1.0.0

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova committed:
ADR003: Add pros and cons of the options

Describe pros of developing TUF 1.0.0 in a subdirectory
of the current implementation against the rest of the options.

Signed-off-by: Teodora Sechkova <tsechkova@vmware.com>

sechkova commented:
Updated with your suggestions.

lukpueh commented:
> Updated with your suggestions.

Thanks! Could you please also resolve the conflict in `docs/adr/index.md`?

sechkova commented:
> > Updated with your suggestions.
> 
> Thanks! Could you please also resolve the conflict in `docs/adr/index.md`?

oops .. I forgot .. here it is

lukpueh commented:
picobello!
